### ISSUE TITLE
Subtraction gives incorrect result with zeek tsv formatted input

### ISSUE THREAD

**Issue by ethack** [brimdata/zed#2772]
Subtracting two fields in a `put` with tsv formatted input gives an incorrect result.

```
$ zq -version
Version: v0.29.0
```

Here's a minimal example to reproduce.

```bash
$ cat << EOF | zq -f table 'put diff=(orig_bytes-resp_bytes)' -
#separator \x09
#set_separator  ,
#empty_field    (empty)
#unset_field    -
#fields orig_bytes      resp_bytes
#types  count   count
61      81
EOF
```
Result:
```
ORIG_BYTES RESP_BYTES DIFF
61         81         18446744073709551596
```

This error seems to be specific to Zeek TSV input format. The same example in JSON format works correctly.

```bash
$ echo '{"orig_bytes":61,"resp_bytes":81}' | zq -f table 'put diff=(orig_bytes-resp_bytes)' -                        
```
Result:
```
ORIG_BYTES RESP_BYTES DIFF
61         81         -20
```

I should note that I discovered this issue while processing Zeek logs captured from a real sensor on real world network traffic. So the issue is present there too, and I can provide larger Zeek logs if needed.

My guess is that the types of the bytes when read in are `uint`s and subtracting them is doing unsigned arithmetic. Here's a link to a Go playground that gives the exact same result: `18446744073709551596`.

https://play.golang.org/p/D9b6Fv8_47G

Edit: Since Github doesn't seem to preserve tabs, here is the minimal example file.
[conn-minimal.log](https://github.com/brimdata/zed/files/6548628/conn-minimal.log)

**Comment by philrz** [comment]
@ethack: Thanks for the detailed bug report! indeed, your analysis all seems correct. I'll take up the topic with the Dev team.

You're right that the Zeek `count` type gets internally treated as an unsigned integer after being read in by `zq`, since the [Zeek docs](https://docs.zeek.org/en/current/script-reference/types.html#type-count) do say it's an unsigned int. I'm not sure if you've stumbled across it yet, but since you're a Zeek user you might be interested in the doc at [https://github.com/brimdata/zed/blob/zed-zeek-updates/zeek/type-compat.md](https://github.com/brimdata/zed/blob/zed-zeek-updates/zeek/type-compat.md) that describes the mappings between the two type systems. (That version of the doc happens to be in a branch associated with a PR I've got under review at the moment because we're checking in changes in prep for an upcoming release. But since any fixes to to your issue are going to land in a future release anyway, I figure I might as well point you at the latest stuff. But in the remainder of this comment I'll focus on how things are at the last GA `zq` tagged `v0.29.0`, since I assume that's what you're running.)

You can see this clearly by having `zq` read the Zeek TSV file and print it back out as ZSON, as it shows the explicit `(uint64)` [type decorator]($ zq -f zson 'put diff=(orig_bytes:int64-resp_bytes)' example.log  {     orig_bytes: 61 (uint64),     resp_bytes: 81 (uint64),     diff: -20 } (=0)).

```
$ zq -version
Version: v0.29.0

$ cat example.log 
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#fields	orig_bytes	resp_bytes
#types	count	count
61	81

$ zq -f zson example.log 
{
    orig_bytes: 61 (uint64),
    resp_bytes: 81 (uint64)
} (=0)
```

Looking at your JSON example, there we see no explicit type decorator, which is indicative that it's just inferring the type to be a signed integer (specifically `int64` per the note about "implied-type values" in the ZSON spec section on [Primitive Values](https://github.com/brimdata/zed/blob/main/docs/formats/zson.md#33-primitive-values)). This makes sense since the JSON gives no typing info that would communicate whether or not the field is intended to be treated as signed. (BTW, the thing you see here with the unnecessary use of scientific notation has already been fixed.)

```
$ echo '{"orig_bytes":61,"resp_bytes":81}' | zq -f zson -
{
    orig_bytes: 6.1e+01,
    resp_bytes: 8.1e+01
} (=0)
```

While you're waiting on a fix, the good news is that I can offer you a short-term workaround by using type casting to treat one of the values as signed.

```
$ zq -f zson 'put diff=(orig_bytes:int64-resp_bytes)' example.log 
{
    orig_bytes: 61 (uint64),
    resp_bytes: 81 (uint64),
    diff: -20
} (=0)
```

(Heads-up that the casting notation will be changing in the next release. With current head of `main` of the Zed repo, this would look like `zq -f zson 'put diff:=(int64(orig_bytes)-resp_bytes)' example.log`).

**Comment by ethack** [comment]
Thank you very much for the explanation and the workaround! I didn't know that typecasting was possible and that will work nicely for me.

I know the struggles of inferring types in the different formats of Zeek logs from another project I work on. Ironically, it's usually JSON that usually gives more trouble because of the lack of explicit types, but in this case it actually worked as an advantage :)

**Comment by philrz** [comment]
@ethack: I did a bit more hacking with this and also talked it over with the Dev team. In brief, it seems unlikely we'll change the default treatment here. I can actually see it both ways, since I totally empathize with how at first the result appears incorrect, but it seems to ultimately become a trade-off as to where you want to be exposed to potential incorrectness. Therefore we'd like to put enough controls in the hands of the user to make it do what they really want, and the casting workaround is one such way. There are others I'll get to below, and there may be more ideas for the future.

First, FWIW, I've confirmed our tooling is not alone in this treatment. Within Zeek scripting itself, using their [`count` type](https://docs.zeek.org/en/master/script-reference/types.html#type-count) (which is an unsigned 64-bit integer, hence equivalent to Zed's `uitn64`), the result is the same.

```
$ cat countmath.zeek 
module Hello;
 
export {
    redef enum Log::ID += { LOG };
 
    type Info: record {
        orig_bytes:    count &log;
        resp_bytes:    count &log;
        diff_bytes:    count &log;
        };
    }
     
event zeek_init()
    {
    local orig: count = 61;
    local resp: count = 81;
    local diff: count = orig - resp;
    Log::create_stream(Hello::LOG, [$columns=Hello::Info, $path="hello"]);
    Log::write( Hello::LOG, [$orig_bytes=orig,
                             $resp_bytes=resp,
                             $diff_bytes=diff]);
    }

$ cat hello.log 
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#path	hello
#open	2021-05-27-10-56-15
#fields	orig_bytes	resp_bytes	diff_bytes
#types	count	count	count
61	81	18446744073709551596
#close	2021-05-27-10-56-15
```

MySQL actually refuses to do the computation at all and returns an error. Since Zed is designed to sometimes work on streams, I suspect we'd not want to fail completely like this, but perhaps it could catch what happened and still continue, but return a warning? Such a warning could perhaps point the user toward casting as a possible workaround (and we could cover this specific case in the casting docs.)

```
mysql> CREATE DATABASE mytest;
Query OK, 1 row affected (0.00 sec)

mysql> USE mytest;
Database changed

mysql> CREATE TABLE difftest (
    ->   orig_bytes INT UNSIGNED NOT NULL, 
    ->   resp_bytes INT UNSIGNED NOT NULL
    -> ); 
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO difftest () VALUES(61, 81);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT orig_bytes,resp_bytes,(orig_bytes-resp_bytes) AS result FROM difftest;
ERROR 1690 (22003): BIGINT UNSIGNED value is out of range in '(`mytest`.`difftest`.`orig_bytes` - `mytest`.`difftest`.`resp_bytes`)'

mysql> SELECT orig_bytes,resp_bytes,(cast(orig_bytes as signed)-cast(resp_bytes as signed)) AS result FROM difftest;
+------------+------------+--------+
| orig_bytes | resp_bytes | result |
+------------+------------+--------+
|         61 |         81 |    -20 |
+------------+------------+--------+
1 row in set (0.00 sec)
```

It would be technically feasible for Zed to just change its default treatment of Zeek `count` values and store them as `int64` rather than `uint64`, but this is the trade-off I mentioned above. The current treatment leaves the exposure of your diff computation giving a giant, incorrect-looking results like you saw. If we flipped the treatment, now we'd lose the top half of the numeric range, so now what came from Zeek as very large `count` values might be output from Zed as negative numbers. So it seems to come down to which exposure concerns you more as a user.

The casting workaround gives you a runtime way to make the decision, but I respect it might be a burden to have to do that as part of every query that needs it. Instead you might prefer to just have the fields read and stored with the type treatment flipped. It's not well-documented yet, but here's one way to do this in current `zq` at the tip of `main` (i.e. not yet in a GA release):

```
$ cat shaper.zed 
type connbytes={orig_bytes:int64, resp_bytes:int64};
put this:=shape(connbytes)

$ zq -Z -I shaper.zed '| put diff:=orig_bytes-resp_bytes' example.log 
{
    orig_bytes: 61,
    resp_bytes: 81,
    diff: -20
}
```

Of course, that's just laser-focused on your example log record that has only those to fields. A place where this "shaper" concept is typically used more broadly is when importing JSON, since it's so minimally typed. You already picked up on the fact that reading "unshaped" JSON with `zq` was doing more what you expected since we had no choice but to assume the numeric values might be signed and as large as possible, hence stored them as `int64`. But JSON's minimal typing causes all kinds of other hazards with Zeek data, since timestamps and IP addresses both become strings, there's no way to tell if a JSON array started life as a Zeek `set`, etc. So we publish a reference shaper config for Zeek JSON at [https://github.com/brimdata/zed/blob/main/zeek/shaper.zed](https://github.com/brimdata/zed/blob/main/zeek/shaper.zed) that restores the correct data types for all the fields known to be generated by an out-of-the-box open source Zeek install. If one wanted to make a selective or exhaustive change to how Zeek `count` values get treated at read time, they could tweak that shaper config to have fields that are currently marked as `uint64` stored as `int64`, then apply that shaper config similar to what I did above.

I still find that a little unwieldy since as soon as you start customizing your Zeek with additional fields generated by Zeek packages, or new fields appear when you upgrade to new Zeek versions, now you'd need to keep that config up to date if you want it to remain 100% correct. That seems an appropriate cost to pay if someone's got all their input data as JSON, since so much of it needs fixing regardless. But if starting from Zeek TSV where 99% of the rich typing can alreadyd be derived accurately, but you might just want to flip the behavior for this one `count` type, it seems it would be handy to have a way to apply this kind of config _per-type_ rather than _per-named-field_. That's still a topic being discussed on this end.

I'd be interested to hear your reactions as a user to any of the above.

**Comment by ethack** [comment]
I really appreciate the response and the surrounding discussion. I'll have to give this some more thought, but here is my first reaction.

I completely understand the viewpoint that the math on unsigned ints isn't actually the wrong thing to do as those are the native datatypes from Zeek. And if I were in a programming mindset at the time I would have realized that typecasting would be the way to go and would have looked for that option in the documentation.

I would call myself a programmer, though it is not my day job. I have a computer science degree and experience with signed/unsigned math in both C++ and Go. So I certainly know how to think in code, but, even with that background, when I'm using a tool like `zq` my mindset isn't were I even realize that datatypes and signed/unsigned arithmetic is something I should be concerned about. This was actually my second look at this problem with `zq` as the first time when I was getting unexpected results I just reached for a different tool and got the result I wanted.

For reference, what I was attempting to do was calculate a [producer-consumer ratio calculation](https://github.com/ThreatHuntingProject/ThreatHunting/blob/master/hunts/analyze_producer_consumer_ratio.md), which can range from -1.0 to 1.0. With typecasting, it works quite well with `zq` and I wouldn't be surprised if others discover this use case as well. Here's a full example:

```
$ cat conn.log | zq -f table 'sum_orig=sum(orig_bytes:int64),sum_resp=sum(resp_bytes:int64) by src=id.orig_h,dst=id.resp_h,port=id.resp_p | put pcr=(sum_orig-sum_resp)/(sum_orig+sum_resp+0.001)' -
```

So for me, with a programming background, if I were given hints that I need to pay attention to the underlying datatype and shown how to typecast that would be enough for me. But I'm not sure if someone with a non-programming background would have the same experience.

These are the places in the documentation I'd expect some reference to datatypes in expressions:
- https://github.com/brimdata/zed/blob/main/docs/language/expressions/README.md
- https://github.com/brimdata/zed/blob/main/docs/language/operators/README.md#put

I also understand why Zeek TSV and JSON behave differently. But as a user I value consistency between formats over technical correctness for extreme cases. One of the great benefits of `zq` for me is how it can handle both formats without needing to change my workflow or parse them differently. (Switching between `zeek-cut` and `jq` is painful :smile:) If I have to worry about the differences between formats again then `zq` loses some of its utility for me. I like the idea of the reference shaper file for JSON that you have as I think that would bring consistency for most cases. Personally, I'd want that to be the default for JSON and not have to use `zq -I shaper.zed`. But as long as the documentation is there to inform me this is a thing I might need, I can set it up at install time and make an alias and not have to worry about it again.

**Comment by philrz** [comment]
@ethack: Thanks for sharing the thoughts. It's spawned some nice discussions on this end. In this regard, we have created two issues for our backlog (#2777, #2778) that will hopefully help future generations not struggle as much if they bump into this.

Your thoughts on where the better coverage should be in the docs is spot-on. We're due to make some updates to those, so we'll definitely make sure there's better coverage as part of those updates. In addition, the topic of whether or not users should have to think like programmers is something we've debated as well. We've had this philosophy we call "the gentle slope of Zed", i.e. that users who _don't_ think like programmers should be able to look at the Search bar in Brim more like a web search and not have to worry about things like data types and overflow if all they want to do is simple searches, but that if they _do_ want to start using expressions/analytics, we'd want to advise them early that they'll probably want to review some of the relevant docs so they don't run into surprises, such as happened to you. So we're now thinking of the best way to point that out to users early without scaring off the ones who could still be content with never having to learn it if it's not required for what they want to accomplish.

Your point about "consistency between formats" is certainly valid, but we have aspirations to make `zq` applicable well beyond just Zeek, so it seems doubtful that we'd "lock in" any one shaper config out-of-the-box. In this regard your proposed approach of using a shell alias is probably the best approach for now, though we do have an issue #1059 to allow the Zed tooling to have its own way for persisting common default configs. Even if `zq` is not just for Zeek, I agree that it's essential to make sure Zeek users know the shaper option exists, and I have some Zeek-related docs to update as tracked in #2594 so I've made a note-to-self to emphasize this point.

As I think our future work is now tracked in other granular issues, I'd be inclined to close this specific issue and let the others live on, but I think your feedback here will live on as a handy reference to think about how different kinds of users might approach the tools. I'll hold this one open a bit in the event you have any other thoughts to share before I close. Thanks for all the input!
